tmpfs/tmpfs.go

```diff
@@ -0,0 +1,120 @@
+package tmpfs
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"gopkg.in/src-d/go-billy.v2"
+	"gopkg.in/src-d/go-billy.v2/subdirfs"
+)
+
+type tmpFs struct {
+	fs        billy.Filesystem
+	tmp       billy.Filesystem
+	tempFiles map[string]bool
+}
+
+const defaultCreateMode = 0666
+
+// New creates a new filesystem wrapping up 'fs' and using 'tmp' for temporary
+// files. Any file created with TempFile will be created in 'tmp'. It will be
+// moved to 'fs' if Rename is called on it.
+//
+// This is particularly useful to provide TempFile for filesystems that do not
+// support it or where there is a performance penalty in doing so.
+func New(fs billy.Filesystem, tmp billy.Filesystem) billy.Filesystem {
+	return &tmpFs{
+		fs:        fs,
+		tmp:       tmp,
+		tempFiles: map[string]bool{},
+	}
+}
+
+func (t *tmpFs) Create(path string) (billy.File, error) {
+	return t.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, defaultCreateMode)
+}
+
+func (t *tmpFs) Open(path string) (billy.File, error) {
+	return t.OpenFile(path, os.O_RDONLY, 0)
+
+}
+
+func (t *tmpFs) OpenFile(p string, flag int, mode os.FileMode) (
+	billy.File, error) {
+	if t.isTmpFile(p) {
+		return t.tmp.OpenFile(p, flag, mode)
+	}
+
+	return t.fs.OpenFile(p, flag, mode)
+}
+
+func (t *tmpFs) ReadDir(p string) ([]billy.FileInfo, error) {
+	return t.fs.ReadDir(p)
+}
+
+func (t *tmpFs) Join(elem ...string) string {
+	return t.fs.Join(elem...)
+}
+
+func (t *tmpFs) Dir(p string) billy.Filesystem {
+	return subdirfs.New(t, p)
+}
+
+func (t *tmpFs) MkdirAll(filename string, perm os.FileMode) error {
+	return t.fs.MkdirAll(filename, perm)
+}
+
+func (t *tmpFs) Base() string {
+	return t.fs.Base()
+}
+
+func (t *tmpFs) TempFile(dir string, prefix string) (billy.File, error) {
+	tmpFile, err := t.tmp.TempFile(dir, prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	t.tempFiles[tmpFile.Filename()] = true
+
+	return tmpFile, nil
+}
+
+func (t *tmpFs) Rename(from, to string) error {
+	if t.isTmpFile(to) {
+		return errors.New("cannot rename to temp file")
+	}
+
+	if t.isTmpFile(from) {
+		err := billy.CopyPath(t.tmp, t.fs, from, to)
+		if err != nil {
+			return err
+		}
+
+		return t.tmp.Remove(from)
+	}
+
+	return t.fs.Rename(from, to)
+}
+
+func (t *tmpFs) Remove(path string) error {
```

Comment:
@alcortesm I think it's documented in the interface? In any case, it replicates `os.Remove` behavior.
:+1: to remove it from `tmpFiles` map.